nemo/collections/tts/models/radtts.py

```diff
@@ -0,0 +1,419 @@
+from nemo.core.classes import Exportable
+from nemo.utils import logging
+
+torch.cuda.empty_cache()
+
+
+class RadTTSModel(SpectrogramGenerator, Exportable):
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        if isinstance(cfg, dict):
+            cfg = OmegaConf.create(cfg)
+
+        self._setup_tokenizer(cfg.validation_ds.dataset)
+
+        assert self.tokenizer is not None
+
+        self.tokenizer_pad = self.tokenizer.pad
+        self.tokenizer_unk = self.tokenizer.oov
+
+        self.text_tokenizer_pad_id = None
+        self.tokens = None
+
+        super().__init__(cfg=cfg, trainer=trainer)
+        self.feat_loss_weight = 1.0
+        self.model_config = cfg.modelConfig
+        self.train_config = cfg.trainerConfig
+        self.optim = cfg.optim
+        self.criterion = RADTTSLoss(
+            self.train_config.sigma,
+            self.model_config.n_group_size,
+            self.model_config.dur_model_config,
+            self.model_config.f0_model_config,
+            self.model_config.energy_model_config,
+            vpred_model_config=self.model_config.v_model_config,
+            loss_weights=self.train_config.loss_weights,
+        )
+
+        self.attention_kl_loss = AttentionBinarizationLoss()
+        self.model = instantiate(cfg.modelConfig)
+        self._parser = None
+        self._tb_logger = None
+        self.cfg = cfg
+        self.log_train_images = False
+
+    def batch_dict(self, batch_data):
+        batch_data_dict = {
+            "audio": batch_data[0],
+            "audio_lens": batch_data[1],
+            "text": batch_data[2],
+            "text_lens": batch_data[3],
+            "log_mel": batch_data[4],
+            "log_mel_lens": batch_data[5],
+            "align_prior_matrix": batch_data[6],
+            "pitch": batch_data[7],
+            "pitch_lens": batch_data[8],
+            "voiced_mask": batch_data[9],
+            "p_voiced": batch_data[10],
+            "energy": batch_data[11],
+            "energy_lens": batch_data[12],
+        }
+        # "speaker_id": batch_data[13],}
+        return batch_data_dict
+
+    def forward(self, batch):
+        batch = self.batch_dict(batch)
+        mel = batch['log_mel']
+        speaker_ids = batch['speaker_id']
+        text = batch['text']
+        in_lens = batch['text_lens']
+        out_lens = batch['log_mel_lens']
+        attn_prior = batch['align_prior_matrix']
+        aug_idxs = None
+        f0 = batch['pitch']
+        voiced_mask = batch['voiced_mask']
+        p_voiced = batch['p_voiced']
+        energy_avg = batch['energy']
+        if (
+            self.train_config.binarization_start_iter >= 0
+            and self.global_step >= self.train_config.binarization_start_iter
+        ):
+            # binarization training phase
+            binarize = True
+        else:
+            # no binarization, soft-only
+            binarize = False
```

Comment:
Why is this done in forward, training_step, and validation_step?

Reply:
removed the duplicate.